amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+
+#[cfg(all(feature = "backtrace", feature = "std"))]
+extern crate backtrace;
+
+use self::internal::Backtrace;
+use std::borrow::Cow;
+use std::error;
+use std::fmt;
+use std::result;
+
+/// Internal parts of `Error`.
+enum ErrorImpl {
+    Message(Cow<'static, str>),
+    Error(Box<dyn error::Error + Send + Sync>),
+}
+
+impl fmt::Debug for ErrorImpl {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ErrorImpl::Message(ref message) => fmt::Debug::fmt(message, fmt),
+            ErrorImpl::Error(ref e) => fmt::Debug::fmt(e, fmt),
+        }
+    }
+}
+
+impl From<&'static str> for ErrorImpl {
+    fn from(message: &'static str) -> Self {
+        ErrorImpl::Message(Cow::from(message))
+    }
+}
+
+impl From<String> for ErrorImpl {
+    fn from(message: String) -> Self {
+        ErrorImpl::Message(Cow::from(message))
+    }
+}
+
+impl From<Box<dyn error::Error + Send + Sync>> for ErrorImpl {
+    fn from(error: Box<dyn error::Error + Send + Sync>) -> Self {
+        ErrorImpl::Error(error)
+    }
+}
+
+/// The error type used by Amethyst.
+///
+/// Wraps error diagnostics like messages and other errors, and keeps track of causal chains and
+/// backtraces.
+pub struct Error {
+    error_impl: ErrorImpl,
+    source: Option<Box<Error>>,
```

Comment:
Can we call this `cause`?

Reply:
We could... but the rfc uses `source` since `cause` is occupied. It's not a big deal though since I think we'll build done kind of error handling abstraction in utils which will most likely be the only one using this right now.